### ISSUE TITLE
Instance manager stuck executing fencing off

### ISSUE THREAD

**Issue by gbartolini** [cloudnative-pg/cloudnative-pg#2859]
See the attached log for details. I was able to get here by creating a cluster reusing an existing WAL archive (my bad). The cluster was complaining that WAL archiving was failing, but the archive contained history files for timelines 2 and 3 (the primary was on 1). This left the standby hanging.

At the same time a volume snapshot backup request started, asking to fence the instance and when completed to unfence it.

I attach to this is issue the full output of the log after running `kill -ABRT 1`. This is a relevant excerpt from it:

```
goroutine 47 [sleep]:
runtime.gopark(0x223bd24e19cd5?, 0x10be5d6?, 0x0?, 0xf2?, 0x0?)
        /opt/hostedtoolcache/go/1.21.1/x64/src/runtime/proc.go:398 +0xce fp=0xc0008e59c0 sp=0xc0008e59a0 pc=0x43e38e
time.Sleep(0x12a05f200)
        /opt/hostedtoolcache/go/1.21.1/x64/src/runtime/time.go:195 +0x125 fp=0xc0008e5a00 sp=0xc0008e59c0 pc=0x46e6e5
k8s.io/apimachinery/pkg/util/wait.ExponentialBackoff({0x12a05f200, 0x0, 0x0, 0x7fffed90, 0x0}, 0x43e38e?)
        pkg/mod/k8s.io/apimachinery@v0.28.1/pkg/util/wait/backoff.go:467 +0x45 fp=0xc0008e5a20 sp=0xc0008e5a00 pc=0x10bc905
k8s.io/client-go/util/retry.OnError({0x12a05f200, 0x0, 0x0, 0x7fffffff, 0x0}, 0xc000980301?, 0xc0002fbb00?)
        pkg/mod/k8s.io/client-go@v0.28.1/util/retry/util.go:50 +0xa5 fp=0xc0008e5aa0 sp=0xc0008e5a20 pc=0x18c3825
github.com/cloudnative-pg/cloudnative-pg/pkg/management/postgres.(*Instance).waitForInstanceRestarted(0xc000e66420?, {0x1b?, 0x0?, 0x317a240?})
        pkg/management/postgres/instance.go:1009 +0x7e fp=0xc0008e5b10 sp=0xc0008e5aa0 pc=0x192cc3e
github.com/cloudnative-pg/cloudnative-pg/pkg/management/postgres.(*Instance).RequestAndWaitFencingOff(0xc00030e000)
        pkg/management/postgres/instance.go:987 +0xa6 fp=0xc0008e5ba0 sp=0xc0008e5b10 pc=0x192ca86
github.com/cloudnative-pg/cloudnative-pg/internal/management/controller.(*InstanceReconciler).reconcileFencing(0xc000267400, 0x21d34f8?)
        internal/management/controller/instance_controller.go:310 +0x5b fp=0xc0008e5bf0 sp=0xc0008e5ba0 pc=0x19b125b
github.com/cloudnative-pg/cloudnative-pg/internal/management/controller.(*InstanceReconciler).Reconcile(0xc000267400, {0x21d34f8?, 0xc001182a20?}, {{{0xc001182a20?, 0x0?}, {0xc000986d20?, 0x4108e5?}}})
        internal/management/controller/instance_controller.go:155 +0x453 fp=0xc0008e5cd0 sp=0xc0008e5bf0 pc=0x19b06f3
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x21d34f8?, {0x21d34f8?, 0xc001182a20?}, {{{0xc000238dd6?, 0x1ad1b00?}, {0xc000e660a0?, 0x21c2670?}}})
        pkg/mod/sigs.k8s.io/controller-runtime@v0.15.1/pkg/internal/controller/controller.go:118 +0xb7 fp=0xc0008e5d70 sp=0xc0008e5cd0 pc=0x15b30f7
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc00063c640, {0x21d3530, 0xc00061e7d0}, {0x1cc1260?, 0xc0010200e0?})
        pkg/mod/sigs.k8s.io/controller-runtime@v0.15.1/pkg/internal/controller/controller.go:314 +0x365 fp=0xc0008e5ee0 sp=0xc0008e5d70 pc=0x15b4f45
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc00063c640, {0x21d3530, 0xc00061e7d0})
        pkg/mod/sigs.k8s.io/controller-runtime@v0.15.1/pkg/internal/controller/controller.go:265 +0x1c9 fp=0xc0008e5f80 sp=0xc0008e5ee0 pc=0x15b4789
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
        pkg/mod/sigs.k8s.io/controller-runtime@v0.15.1/pkg/internal/controller/controller.go:226 +0x79 fp=0xc0008e5fe0 sp=0xc0008e5f80 pc=0x15b4219
runtime.goexit()
        /opt/hostedtoolcache/go/1.21.1/x64/src/runtime/asm_amd64.s:1650 +0x1 fp=0xc0008e5fe8 sp=0xc0008e5fe0 pc=0x4715a1
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2 in goroutine 81
        pkg/mod/sigs.k8s.io/controller-runtime@v0.15.1/pkg/internal/controller/controller.go:222 +0x565
```

[abrt.log](https://github.com/cloudnative-pg/cloudnative-pg/files/12713479/abrt.log)

**Comment by jsilvela** [comment]
I think this is related to https://github.com/cloudnative-pg/cloudnative-pg/issues/2853
Also, we were not de-bouncing the Unfencing.

**Comment by jsilvela** [comment]
When unfencing, we may error out if the pod was not fenced. Fixed in the ticket I mentioned, by adding a de-bouncing predicate to the unfencing function.

``` go
        ...
	); err != nil && !errors.Is(err, utils.ErrorServerAlreadyUnfenced) {
 		return err
 	}
```

**Comment by jsilvela** [comment]
I've looked into the log in more detail.
I don't think this is a problem with the fencing off itself.
As you pointed out, the instance was failing because of `requested timeline 3 is not a child of this server's history`.

Before #2853 was fixed, what could happen is that if the snapshot backup taken did not have a proper "Completed" status, the snapshot reconciler would immediately issue a fencing request, even if the snapshots were done already.

In the volume snapshot reconciler then, we would:

1. Request the pod to be fenced
2. Consider the pod is fenced when it stops being ready (see code below) --- since the pod was failing due to the timeline issue, it was definitely not ready, so seen as "Fenced".
3. Find there were volume snapshots already, so, backup "done"
4. Ask for the pod to be unfenced, for which we need to instance to be ready (see second code excerpt)

Now, I think that considering a pod that is failing to be "fenced" is a head-scratcher.
We might want to make that more explicit.

Code excerpt 1.

``` go
waitForPodToBeFenced(
	ctx context.Context,
	targetPod *corev1.Pod,
) (*ctrl.Result, error) {
	contextLogger := log.FromContext(ctx)

	var pod corev1.Pod
	err := se.cli.Get(ctx, types.NamespacedName{Name: targetPod.Name, Namespace: targetPod.Namespace}, &pod)
	if err != nil {
		return nil, err
	}
	ready := utils.IsPodReady(pod)
	if ready {
		contextLogger.Info("Waiting for target Pod to not be ready, retrying", "podName", targetPod.Name)
		return &ctrl.Result{RequeueAfter: 10 * time.Second}, nil
	}
```

Code excerpt 2.

``` go
	instance.requestFencingOff()
	err := instance.waitForInstanceRestarted(now)
```
